File: SonicVale/app/core/tts_capabilities.py
"""Pure model/voice capability resolution shared by request adapters and the UI API."""

def cosyvoice_instruction_mode(model, params, voice_name=None):
    """Resolve capabilities per voice; most v3 system voices have no Instruct."""
    configured = str(params.get("instruction_mode") or "auto").strip().lower()
    if configured in {"mapped", "none"}:
        return configured
    if params.get("supports_instruction") is False:
        return "mapped"
    model = (model or "").lower()
    if model.startswith(("cosyvoice-v1", "cosyvoice-v2")):
        return "mapped"
    if model.startswith(("cosyvoice-v3.5-plus", "cosyvoice-v3.5-flash")):
        return "native"
    if model.startswith(("cosyvoice-v3-flash", "cosyvoice-v3-plus")):
        voice = voice_name or params.get("voice") or "longanyang"
        if voice in {"longanyang", "longanhuan", "longhuhu_v3"}:
            return "structured"
        if voice.startswith("cosyvoice-") and model.startswith("cosyvoice-v3-flash"):
            return "native"
        # longanhuan_v3 accepts dialect instructions only, not free acting notes.
        # Other documented v3 system voices support prosody but not Instruct.
        return "mapped"
    if configured in {"native", "structured"}:
        return configured
    return "native" if params.get("supports_instruction") is True else "none"


def http_instruction_field(model, params):
    """Return the provider-specific instruction path for non-CosyVoice HTTP adapters."""
    configured = params.get("instruction_field")
    if configured is False or params.get("supports_instruction") is False:
        return None
    if isinstance(configured, str) and configured.strip():
        return configured.strip()
    model = (model or "").lower()
    if (model or "").lower() in {"qwen-audio-3.0-tts-plus", "qwen-audio-3.0-tts-flash"}:
        return "input.instruction"
    if "qwen" in model and "instruct" in model:
        return "input.instructions"
    if model.startswith("gpt-4o") and "tts" in model:
        return "instructions"
    return None

File: sonicvale-front/src/utils/ttsCapabilities.js
export const QWEN_DRAMA_MODEL = 'qwen-audio-3.0-tts-plus'

// Historical Demo recordings retain their original model/voice identity.
export const QWEN_DRAMA_VOICES = [
  { voice: 'Moon', label: '月白 · 青年男声', role: '都市男角', sample: 'moon' },
  { voice: 'Maia', label: '四月 · 知性女声', role: '都市女角', sample: 'maia' },
  { voice: 'Vincent', label: '田叔 · 沙哑男声', role: '旁白', sample: 'vincent' },
]

export const QWEN_AUDIO_VOICES = {
  'qwen-audio-3.0-tts-plus': [
    { voice: 'longanlingxin', label: '龙安灵心 · 温暖女声', role: '都市女角', sampleFile: 'free-auditions/qwen-audio-plus-lingxin.mp3' },
    { voice: 'longanlufeng', label: '龙安鲁风 · 明亮男声', role: '都市男角', sampleFile: 'free-auditions/qwen-audio-plus-lufeng.mp3' },
  ],
  'qwen-audio-3.0-tts-flash': [
    { voice: 'longanfengyue', label: '龙安风悦 · 自然女声', role: '都市女角', sampleFile: 'free-auditions/qwen-audio-flash-fengyue.mp3' },
    { voice: 'longanhuan_v3.6', label: '龙安欢 · 青年女声', role: '都市女角' },
    { voice: 'longchuanshu_v3.6', label: '龙川叔 · 川普男声', role: '成熟角色' },
  ],
}

export function qwenDramaVoices(model) {
  return QWEN_AUDIO_VOICES[String(model || '').toLowerCase()] || QWEN_DRAMA_VOICES
}

export function isQwenDramaModel(model) {
  return !!QWEN_AUDIO_VOICES[String(model || '').toLowerCase()] || /^qwen3-tts-instruct-flash(?:-\d{4}-\d{2}-\d{2})?$/.test(String(model || '').toLowerCase())
}

// Mirror ConfigurableCloudTTSEngine; CosyVoice capability belongs to the voice,
// so the provider card describes its configured default voice only.
export function ttsCapability(provider, voiceName) {
  if (provider?.provider_type === 'edge') return { label: '基础韵律', type: 'info', mode: 'mapped' }
  let params = {}
  try { params = typeof provider?.custom_params === 'object' ? provider.custom_params || {} : JSON.parse(provider?.custom_params || '{}') } catch { params = {} }
  if (!params || typeof params !== 'object' || Array.isArray(params)) params = {}
  const model = String(provider?.model || '').toLowerCase()
  const configured = params.instruction_mode
  let mode = 'none'
  if (model.startsWith('cosyvoice')) {
    if (['mapped', 'none'].includes(configured)) mode = configured
    else if (params.supports_instruction === false || /^cosyvoice-v[12]/.test(model)) mode = 'mapped'
    else if (model.startsWith('cosyvoice-v3.5')) mode = 'native'
    else if (/^cosyvoice-v3-(flash|plus)/.test(model)) {
      const voice = voiceName || params.voice || 'longanyang'
      mode = ['longanyang', 'longanhuan', 'longhuhu_v3'].includes(voice) ? 'structured'
        : model.startsWith('cosyvoice-v3-flash') && voice.startsWith('cosyvoice-') ? 'native' : 'mapped'
    } else if (['native', 'structured'].includes(configured)) mode = configured
    else if (params.supports_instruction === true) mode = 'native'
  } else if (JSON.stringify(params.payload || params.body || {}).includes('{{instruction}}')) mode = 'native'
  else if (params.supports_instruction !== false && params.instruction_field !== false && (
    params.instruction_field || !!QWEN_AUDIO_VOICES[model] || (model.includes('qwen') && model.includes('instruct')) || (model.startsWith('gpt-4o') && model.includes('tts'))
  )) mode = 'native'
  return {
    native: { label: '原生表演指令', type: 'success', mode },
    structured: { label: '固定情感指令', type: 'success', mode },
    mapped: { label: '基础韵律', type: 'warning', mode },
    none: { label: '基础生成', type: 'info', mode },
  }[mode]
}
